metricproviders/kayenta/kayenta.go

```diff
@@ -0,0 +1,271 @@
+		}
+		return "", err
+	}
+
+	data, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return "", err
+	}
+
+	var cc []canaryConfig
+
+	err = json.Unmarshal(data, &cc)
+	if err != nil {
+		return "", err
+	}
+
+	for _, s := range cc {
+		if s.Name == metric.Provider.Kayenta.CanaryConfigName {
+			return s.Id, nil
+		}
+	}
+
+	return "", err
+}
+
+// Run queries kayentd for the metric
+func (p *Provider) Run(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric) v1alpha1.Measurement {
+	startTime := metav1.Now()
+	newMeasurement := v1alpha1.Measurement{
+		StartedAt: &startTime,
+	}
+
+	canaryConfigId, err := getCanaryConfigId(metric, p)
+	if err != nil {
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+	}
+
+	jobURL := fmt.Sprintf(jobURLFormat, metric.Provider.Kayenta.Address, canaryConfigId, metric.Provider.Kayenta.Application, metric.Provider.Kayenta.MetricsAccountName, metric.Provider.Kayenta.ConfigurationAccountName, metric.Provider.Kayenta.StorageAccountName)
+
+	var scopes string
+	for i, s := range metric.Provider.Kayenta.Scopes {
+		name := s.Name
+		controlScope, err := json.Marshal(s.ControlScope)
+		if err != nil {
+			return metricutil.MarkMeasurementError(newMeasurement, err)
+		}
+
+		experimentScope, err := json.Marshal(s.ExperimentScope)
+		if err != nil {
+			return metricutil.MarkMeasurementError(newMeasurement, err)
+		}
+		scopes = scopes + fmt.Sprintf(scopeFormat, name, string(controlScope), string(experimentScope))
+		if i < (len(metric.Provider.Kayenta.Scopes) - 1) {
+			scopes = scopes + ","
+		}
+
+	}
+
+	jobPayLoad := fmt.Sprintf(jobPayloadFormat, scopes, metric.Provider.Kayenta.Threshold.Pass, metric.Provider.Kayenta.Threshold.Marginal)
+
+	jsonValue, err := templateutil.ResolveArgs(jobPayLoad, run.Spec.Args)
+	if err != nil {
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+	}
+
+	response, err := p.client.Post(jobURL, "application/json", bytes.NewBuffer([]byte(jsonValue)))
+	if err != nil || response.Body == nil || response.StatusCode != 200 {
+		if err == nil {
+			err = errors.New("Invalid Response")
+		}
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+	}
+
+	data, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+	}
+	var dat map[string]interface{}
+	if err := json.Unmarshal(data, &dat); err != nil {
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+	}
+	jobId := dat["canaryExecutionId"]
+	m := make(map[string]string)
+	m["canaryExecutionId"] = fmt.Sprintf("%v", jobId)
+	if len(m["canaryExecutionId"]) == 0 {
+
+		return metricutil.MarkMeasurementError(newMeasurement, errors.New("Invalid canaryExecutionId"))
+	}
+	newMeasurement.Metadata = m
+
+	newMeasurement.Phase = v1alpha1.AnalysisPhaseRunning
+
+	resumeTime := metav1.NewTime(time.Now().Add(resumeDelay))
+	newMeasurement.ResumeAt = &resumeTime
+	finishTime := metav1.Now()
+	newMeasurement.FinishedAt = &finishTime
```

Comment:
The run step for Kayenta should not set the finishedAt field. The finishedAt time should only be set when the measurement is finished. In the Kayenta case, the measurement is finished when the returned Kayenta execution is complete and we evaluated the result.